File: src/main/kotlin/hyunsoo/48week/순위.kt
package hyunsoo.`48week`

/**
 *
 * <문제>
 * [순위](https://school.programmers.co.kr/learn/courses/30/lessons/49191)
 *
 * - 아이디어
 *
 * - 트러블 슈팅
 *
 */
class `전현수_순위` {

    fun solution(n: Int, results: Array<IntArray>): Int {

        val resultTable = Array(n + 1) {
            IntArray(n + 1) { INIT }
        }

        results.forEach { result ->
            val (winPos, losePos) = result
            resultTable[winPos][losePos] = WIN
            resultTable[losePos][winPos] = LOSE
        }

        for (k in 1 .. n) {
            for (i in 1 .. n) {
                for (j in 1 .. n) {
                    if (resultTable[i][k] == WIN && resultTable[k][j] == WIN) {
                        resultTable[i][j] = WIN
                        resultTable[j][i] = LOSE
                    }
                }
            }
        }

        return resultTable.drop(1).count { row ->
            row.drop(1).count { it == INIT} <= 1
        }
    }

    companion object {
        const val WIN = 1
        const val LOSE = -1
        const val INIT = 0
    }
}

fun main() {
    전현수_순위().solution(
        5, arrayOf(
            intArrayOf(4,3),
            intArrayOf(4,2),
            intArrayOf(3,2),
            intArrayOf(1,2),
            intArrayOf(2,5),
        )
    ).apply {
        println(this)
    }
}

File: src/main/kotlin/jimin/48week/순위.py
'''
플루이드 워셜로 풀었다(질문 게시판을 봤다)
근데 왜 k, i, j 순으로만 풀어야할까?
i, j ,k 순으로 하니 틀린다!
'''
def solution(n, results):
    players = [[0 for _ in range(n + 1)] for _ in range(n + 1)]
    for i in range(len(results)):
        a, b = results[i]
        players[a][b] = 1
        players[b][a] = -1

    for k in range(1, n + 1):
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                if i != j and players[i][j] == 0:
                    if players[i][k] * players[k][j] > 0:
                        players[i][j] = players[i][k]

    result = 0
    for i in range(1, n + 1):
        num = 0
        for j in range(1, n + 1):
            if players[i][j] == 0:
                num += 1
        if num == 1:
            result += 1


    return result